### PR TITLE
[@types/node] use assertion guards in assert module

### LINKED ISSUE #40415: @types/node Use Typescript 3.7 asserts syntax for node's assert function
About the `asserts` feature:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions
https://github.com/microsoft/TypeScript/pull/32695

Basically I would like to change:
https://github.com/DefinitelyTyped/DefinitelyTyped/blob/9f57df89f30bacee1d46dfbf40a08132b4b059ef/types/node/assert.d.ts#L2

into this:
```ts
function internal(value: any, message?: string | Error): asserts value;
```

Since this TypeScript feature has been released only recently, how should this best be added to the types without breaking compilation on older TypeScript versions?

cc @Flarna @rbuckton @jstewmon (This is obviously not the complete list of authors because I didn't want to ping too many people. While browsing the source these names came up. Please ping anyone who you think should be involved in the discussion)

### PR DESCRIPTION
**Pull request by G-Rath**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] Add or edit tests to reflect the change. (Run with `npm test`.)
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: [assertion functions](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-7.html#assertion-functions)
- [ ] If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`. If for reason the any rule need to be disabled, disable it for that line using `// tslint:disable-next-line [ruleName]` and not for whole package so that the need for disabling can be reviewed.

I've made a start on this, but it was getting a bit tangly so wanted to push it up for the diff & for some guidance on how to handle a few things :)

Closes #40415

Lets try this on for size...

### CODE REVIEW

**Comment by G-Rath** [comment]
If this passes, I'll rebase & adjust the commits, so lets not merge.

I'm interested in the order that `typesVersions` should be in, b/c it seems that `ts3.2` is before `ts3.5`, but when I had `ts3.7` after `ts3.5` it didn't work, so now I have no idea what's going on 🤷‍♂ 

Anyway, @SimonSchick watch this space.

**Comment by G-Rath** [comment]
@typescript-bot yeah yeah I know why. I'll fix it when I'm home

**Comment by SimonSchick** [comment]
@G-Rath you do know you can run those tests locally right?

**Comment by G-Rath** [comment]
@SimonSchick it works 🎉🎉🎉 - the bulk of the tests are passing.

There are at least two failures however, which are weird ones:

```
2> Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/node/v10/ts3.7/node-tests.ts:20:5
2> ERROR: 20:5  expect  TypeScript@next expected type to be:
2>   null | undefined
2> got:
2>   null
2> 
2>     at /home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js:193:19
2>     at Generator.next (<anonymous>)
2>     at fulfilled (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/index.js:6:58)
```

That doesn't seem right; using assert locally the resulting type is `null | undefined`, but for now I'm happy to ignore that in favor of the other issue, which I suspect is going to be a massive pain:

```
Error in adone
Error: /home/travis/build/DefinitelyTyped/DefinitelyTyped/types/adone/test/glosses/std.ts:7:9
ERROR: 7:9  expect  TypeScript@next compile error: 
Assertions require every name in the call target to be declared with an explicit type annotation.
```

`std` re-exports `assert`, and I'm not sure how assertion guards behave in that situation 😬

**Comment by G-Rath** [comment]
> @G-Rath you do know you can run those tests locally right?

You beat me to the punch :joy:

I do indeed, but the tests are actually passing for the most part - the failure is in a single library, and running those tests locally eats up resources like crazy since it's installing so many libraries.

I have tested this out locally against libraries in `node_modules`, which is why the `v10` failure is so weird, but they've definitely now actually passing in general which is great.

**Comment by G-Rath** [comment]
So the `v10` failure has gone away, so now it's just failing on the `adone` types.

I'm going to revist the `v10` failure once I've got the build passing, b/c the change I had to make is not required for the other versions, which is super weird (unless I'm missing something special about v10?).

**Comment by G-Rath** [comment]
Ugh so I believe this is caused by https://github.com/microsoft/TypeScript/issues/35004, and renders `std.assert` completely unusable.

We can fix the test by removing the assertion guard from `function internal`, but it'll actually happen on *any* assert function that has a guard.

I'll play around later with seeing if there's a way to have the types for that library do something different in ts3.7, like not use ts3.7 `assert`, but this is the current blocker from shipping :(

**Comment by G-Rath** [comment]
[`adone` is actually deprecated](https://www.npmjs.com/package/adone), so I'm wondering if we can get away with breaking these types, or just hacking a fix along the lines of just copying the whole assert types from `@node/types` 🤷‍♂

**Comment by G-Rath** [comment]
@SimonSchick

you do know you can run those tests locally right?

Actually running them locally results in an error:

> Error: In package.json: Dependency electron-store not in whitelist.
> If you are depending on another `@types` package, do *not* add it to a `package.json`. Path mapping should make the import work.
> For namespaced dependencies you then have to add a `paths` mapping from `@namespace/library` to `namespace__library` in `tsconfig.json`.

File: types/adone/glosses/std.d.ts
Comment (G-Rath) [inline types/adone/glosses/std.d.ts]: The only dependent for this package on npm is [bloater](https://www.npmjs.com/package/bloater), which just depends on the first 1000 dependencies on npm, and is an attempt to just publish a really large package by the looks of it.

The `adone` npm package itself is deprecated, and the github project, while recently updated ~20 days ago, doesn't seem to have the `std` library anymore.

Of the two maintainers of these types, one no longer exists on github, and the other has no active contributions.

I believe it's acceptable in this case to slightly weaken the types of a single package that doesn't appear to be used anymore, vs not having assertion guards in `@types/node`.

Happy to break this out into another PR if needed.

File: types/node/ts3.7/assert.d.ts
Comment (felixfbecker) [inline types/node/ts3.7/assert.d.ts]: Any reason this is called "internal" and not just `assert`? I wonder if TS auto-import can work better if it is named the way people usually import it...

**Review by felixfbecker — APPROVED** [review]

File: types/node/ts3.7/assert.d.ts
Comment (rbuckton) [reply types/node/ts3.7/assert.d.ts]: This also makes quick info for `assert` in `import assert = require("assert")` appear as `namespace internal (alias)`, which would be undesirable. I agree that this should be changed to `assert`.

File: types/node/package.json
Comment (rbuckton) [inline types/node/package.json]: The 3.5 check should come before the 3.2 check, though that primarily relies on for-in enumeration ordering of the keys of the object. A more reliable mechanism would be to use non-overlapping ranges:

```json
{
  "typesVersions": {
    ">=3.7.0.0-0": { ... },
    ">=3.5.0.0-0 <3.7.0.0-0": { ... },
    ">=3.2.0.0-0 <3.5.0.0-0": { ... }
  }
}
```

**Review by rbuckton — CHANGES_REQUESTED** [review]

**Comment by G-Rath** [comment]
@rbuckton 

> appear as namespace internal (alias), which would be undesirable.

This is the case for `stream` & `events as well - would you like me to make a follow up PR after this one updating those as well?

File: types/node/package.json
Comment (G-Rath) [reply types/node/package.json]: Travis doesn't seem to like that:

```
2> node START
2> node failing:
2> AssertionError [ERR_ASSERTION]: "typesVersions" in '/home/travis/build/DefinitelyTyped/DefinitelyTyped/types/node/package.json' is not set right. Should be: {
2>     ">=3.2.0-0": {
2>         "*": [
2>             "ts3.2/*"
2>         ]
2>     },
2>     ">=3.5.0-0": {
2>         "*": [
2>             "ts3.5/*"
2>         ]
2>     },
2>     ">=3.7.0-0": {
2>         "*": [
2>             "ts3.7/*"
2>         ]
2>     }
2> }
2>     at Object.<anonymous> (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/checks.js:35:20)
2>     at Generator.next (<anonymous>)
2>     at fulfilled (/home/travis/build/DefinitelyTyped/DefinitelyTyped/node_modules/dtslint/bin/checks.js:5:58)
```

**Comment by G-Rath** [comment]
It seems that uuid has been broken in some way, causing this PR to error.

It seems like #43210 is an attempt to fix it, but is currently erroring.

@rbuckton would it be possible to get this PR locked in someway to prevent it being autoclosed?

**Comment by G-Rath** [comment]
@felixfbecker I can't request a re-review from you :(

File: types/node/package.json
Comment (G-Rath) [reply types/node/package.json]: I've reverted this change for now so that the build passes, but adjusted the order in `package.json`.

I've made a [revert commit](https://github.com/DefinitelyTyped/DefinitelyTyped/pull/42786/commits/c6c2a4ce6586a86ada902276e3ec9d05c267d3be) so that you can review the changes, which I'll drop before this branch is merged.

**Review by felixfbecker — APPROVED** [review]
Ship it 🚢 🎉

**Comment by G-Rath** [comment]
@rbuckton any chance of a re-review? this PR is ready to go otherwise :)

**Comment by syabro** [comment]
@rbuckton could you please review it? :)

**Comment by syabro** [comment]
Can it be merged without @rbuckton? I'm not sure he is interested in reviewing it :(

**Review by rbuckton — APPROVED** [review]

**Comment by rbuckton** [comment]
My only comment was about the order in `"typesVersions"`, which seems to have been addressed.

**Comment by G-Rath** [comment]
@sandersn just if it matters @typescript-bot didn't post here about publishing a release of `@types/node@12.12.37` to npm.

**Comment by felixfbecker** [comment]
@G-Rath how can I install these when using TS 3.8.3? If I install `@types/node@13.13.2` I still get the `void` return type on `assert()` 🤔

**Comment by G-Rath** [comment]
@felixfbecker unfortunately right now they're not usable because [the `typeVersions` property is being sorted](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/44117) before publishing.

In order for TypeScript to use the new typings, that property has to have it's properties inverted; you can use [`patch-package`](https://www.npmjs.com/package/patch-package) to correct this until the types publisher is fixed & the types are published.

@sandersn seemed pretty active about it, so  I imagine it'll be fixed "soon" 🙂

**Comment by justinmchase** [comment]
Was the version 12 version not published for any particular reason?

**Comment by ackwell** [comment]
Per the [comment in the other issue](https://github.com/DefinitelyTyped/DefinitelyTyped/issues/44117#issuecomment-626010567), it looks like the above issue has indeed been solved. `12.12.38` and `13.13.5` both contain the non-mangled `typesVersion` property.

That said, it looks like it's only those two versions, other major node version releases i.e. `10.x.y` are stale.

Is there a way a new patch release can be pushed of the appropriate major node versions containing the fix to the property?

**Pull request merged**
